File: module2/class4/example1-m1p4e1-bubble-sort-optimized.cs
/*
 * PROFILING EXAMPLE: Optimized Bubble Sort Performance Solution
 * 
 * This example demonstrates optimized Bubble Sort implementations:
 * - Bubble Sort with early termination optimization
 * - Cocktail Sort (bidirectional bubble sort)
 * - Optimized Bubble Sort with reduced comparisons
 * - Efficient comparison and swap operations
 * 
 * OBJECTIVES:
 * - Demonstrate optimization techniques for Bubble Sort
 * - Show performance improvements through better Bubble Sort variants
 * - Compare inefficient vs optimized Bubble Sort solutions
 * - Identify best practices for Bubble Sort algorithm design
 * - Prepare reflection on algorithm optimization
 * 
 * NOTE: This code demonstrates optimized Bubble Sort implementations.
 * Run this with Visual Studio Profiler in Instrumentation mode
 * to observe performance improvements and optimization patterns.
 */

using System;
using System.Diagnostics;

namespace ProfilingExample
{
    class Program
    {
        // ============================================================================
        // CLASSROOM CONFIGURATION - EASY TO ADJUST FOR DIFFERENT DEMONSTRATIONS
        // ============================================================================

        // Array Configuration
        private const int ARRAY_SIZE = 5000;              // Array size for profiling (same as problem version)
        private const int TEST_ITERATIONS = 3;            // Number of test iterations
        private const int RANDOM_SEED = 42;               // Random seed for reproducible results

        // Performance Tracking
        private static long totalComparisons = 0;
        private static long totalSwaps = 0;

        // ============================================================================

        /*
         * SCENARIO 1: Optimized Bubble Sort Implementation
         * Demonstrates early termination and efficiency improvements
         */

        // OPTIMIZED: Bubble Sort with early termination - O(n²) worst case, O(n) best case
        private static void BubbleSortOptimized(int[] arr)
        {
            int n = arr.Length;
            totalComparisons = 0;
            totalSwaps = 0;

            // OPTIMIZED: Early termination when no swaps occur
            for (int i = 0; i < n - 1; i++)
            {
                bool swapped = false;

                for (int j = 0; j < n - i - 1; j++)
                {
                    totalComparisons++;

                    // OPTIMIZED: Efficient comparison and swap
                    if (arr[j] > arr[j + 1])
                    {
                        // OPTIMIZED: Efficient swap operation
                        Swap(arr, j, j + 1);
                        totalSwaps++;
                        swapped = true;
                    }
                }

                // OPTIMIZED: Early termination if no swaps occurred
                if (!swapped)
                {
                    break;
                }
            }
        }

        // OPTIMIZED: Efficient swap operation
        private static void Swap(int[] arr, int i, int j)
        {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }

        /*
         * SCENARIO 2: Cocktail Sort Implementation
         * Demonstrates bidirectional bubble sort optimization
         */

        // OPTIMIZED: Cocktail Sort (bidirectional bubble sort) - O(n²) worst case, O(n) best case
        private static void CocktailSort(int[] arr)
        {
            int n = arr.Length;
            totalComparisons = 0;
            totalSwaps = 0;
            bool swapped = true;
            int start = 0;
            int end = n - 1;

            while (swapped)
            {
                swapped = false;

                // OPTIMIZED: Forward pass
                for (int i = start; i < end; i++)
                {
                    totalComparisons++;
                    if (arr[i] > arr[i + 1])
                    {
                        Swap(arr, i, i + 1);
                        totalSwaps++;
                        swapped = true;
                    }
                }

                if (!swapped)
                    break;

                swapped = false;
                end--;

                // OPTIMIZED: Backward pass
                for (int i = end; i > start; i--)
                {
                    totalComparisons++;
                    if (arr[i] < arr[i - 1])
                    {
                        Swap(arr, i, i - 1);
                        totalSwaps++;
                        swapped = true;
                    }
                }

                start++;
            }
        }

        /*
         * SCENARIO 3: Optimized Bubble Sort with Reduced Comparisons
         * Demonstrates further optimization techniques
         */

        // OPTIMIZED: Bubble Sort with reduced comparisons and early termination
        private static void BubbleSortReducedComparisons(int[] arr)
        {
            int n = arr.Length;
            totalComparisons = 0;
            totalSwaps = 0;

            for (int i = 0; i < n - 1; i++)
            {
                bool swapped = false;
                int lastSwapIndex = 0;

                for (int j = 0; j < n - i - 1; j++)
                {
                    totalComparisons++;
                    if (arr[j] > arr[j + 1])
                    {
                        Swap(arr, j, j + 1);
                        totalSwaps++;
                        swapped = true;
                        lastSwapIndex = j;
                    }
                }

                // OPTIMIZED: Skip already sorted elements
                if (!swapped)
                    break;

                // OPTIMIZED: Reduce the range for next iteration
                n = lastSwapIndex + 1;
            }
        }

        /*
         * SCENARIO 4: Performance Testing Functions
         */

        private static void TestOptimizedBubbleSort(int iterations)
        {
            Console.WriteLine("=== TESTING OPTIMIZED BUBBLE SORT ===");
            Console.WriteLine("This demonstrates early termination optimization");
            Console.WriteLine($"Array size: {ARRAY_SIZE}");
            Console.WriteLine($"Iterations: {iterations}");
            Console.WriteLine();

            var random = new Random(RANDOM_SEED);
            var stopwatch = new Stopwatch();

            for (int i = 0; i < iterations; i++)
            {
                Console.WriteLine($"Testing Optimized Bubble Sort (iteration {i + 1})...");
                Console.WriteLine("  Creating array with random data...");

                // Create array with random data
                int[] arr = new int[ARRAY_SIZE];
                for (int j = 0; j < ARRAY_SIZE; j++)
                {
                    arr[j] = random.Next(1, 10000);
                }
                Console.WriteLine($"  Array created with {ARRAY_SIZE} random elements");

                // Test optimized Bubble Sort
                Console.WriteLine("  Starting Optimized Bubble Sort...");
                stopwatch.Restart();
                BubbleSortOptimized(arr);
                stopwatch.Stop();
                Console.WriteLine("  Sorting completed!");

                Console.WriteLine($"Optimized Bubble Sort completed:");
                Console.WriteLine($"  Time: {stopwatch.ElapsedMilliseconds} ms");
                Console.WriteLine($"  Comparisons: {totalComparisons:N0}");
                Console.WriteLine($"  Swaps: {totalSwaps:N0}");
                Console.WriteLine($"  Array size: {ARRAY_SIZE}");
                Console.WriteLine();
            }
        }

        private static void TestCocktailSort(int iterations)
        {
            Console.WriteLine("=== TESTING COCKTAIL SORT ===");
            Console.WriteLine("This demonstrates bidirectional bubble sort optimization");
            Console.WriteLine($"Array size: {ARRAY_SIZE}");
            Console.WriteLine($"Iterations: {iterations}");
            Console.WriteLine();

            var random = new Random(RANDOM_SEED);
            var stopwatch = new Stopwatch();

            for (int i = 0; i < iterations; i++)
            {
                Console.WriteLine($"Testing Cocktail Sort (iteration {i + 1})...");
                Console.WriteLine("  Creating array with random data...");

                // Create array with random data
                int[] arr = new int[ARRAY_SIZE];
                for (int j = 0; j < ARRAY_SIZE; j++)
                {
                    arr[j] = random.Next(1, 10000);
                }
                Console.WriteLine($"  Array created with {ARRAY_SIZE} random elements");

                // Test Cocktail Sort
                Console.WriteLine("  Starting Cocktail Sort (bidirectional bubble sort)...");
                stopwatch.Restart();
                CocktailSort(arr);
                stopwatch.Stop();
                Console.WriteLine("  Cocktail Sort completed!");

                Console.WriteLine($"Cocktail Sort completed:");
                Console.WriteLine($"  Time: {stopwatch.ElapsedMilliseconds} ms");
                Console.WriteLine($"  Comparisons: {totalComparisons:N0}");
                Console.WriteLine($"  Swaps: {totalSwaps:N0}");
                Console.WriteLine($"  Array size: {ARRAY_SIZE}");
                Console.WriteLine();
            }
        }

        private static void TestBubbleSortReducedComparisons(int iterations)
        {
            Console.WriteLine("=== TESTING BUBBLE SORT WITH REDUCED COMPARISONS ===");
            Console.WriteLine("This demonstrates further bubble sort optimizations");
            Console.WriteLine($"Array size: {ARRAY_SIZE}");
            Console.WriteLine($"Iterations: {iterations}");
            Console.WriteLine();

            var random = new Random(RANDOM_SEED);
            var stopwatch = new Stopwatch();

            for (int i = 0; i < iterations; i++)
            {
                Console.WriteLine($"Testing Bubble Sort with Reduced Comparisons (iteration {i + 1})...");
                Console.WriteLine("  Creating array with random data...");

                // Create array with random data
                int[] arr = new int[ARRAY_SIZE];
                for (int j = 0; j < ARRAY_SIZE; j++)
                {
                    arr[j] = random.Next(1, 10000);
                }
                Console.WriteLine($"  Array created with {ARRAY_SIZE} random elements");

                // Test Bubble Sort with Reduced Comparisons
                Console.WriteLine("  Starting Bubble Sort with Reduced Comparisons...");
                stopwatch.Restart();
                BubbleSortReducedComparisons(arr);
                stopwatch.Stop();
                Console.WriteLine("  Bubble Sort with Reduced Comparisons completed!");

                Console.WriteLine($"Bubble Sort with Reduced Comparisons completed:");
                Console.WriteLine($"  Time: {stopwatch.ElapsedMilliseconds} ms");
                Console.WriteLine($"  Comparisons: {totalComparisons:N0}");
                Console.WriteLine($"  Swaps: {totalSwaps:N0}");
                Console.WriteLine($"  Array size: {ARRAY_SIZE}");
                Console.WriteLine();
            }
        }

        private static void TestWithDifferentArraySizes()
        {
            Console.WriteLine("=== TESTING OPTIMIZED BUBBLE SORT WITH DIFFERENT ARRAY SIZES ===");
            Console.WriteLine("This demonstrates O(n²) complexity scaling with optimizations");
            Console.WriteLine();

            var random = new Random(RANDOM_SEED);
            var stopwatch = new Stopwatch();

            int[] sizes = { 100, 500, 1000, 2000, 3000 };

            foreach (int size in sizes)
            {
                Console.WriteLine($"Testing with array size: {size}");
                Console.WriteLine($"  Creating array with {size} random elements...");

                // Create array with random data
                int[] arr = new int[size];
                for (int j = 0; j < size; j++)
                {
                    arr[j] = random.Next(1, 10000);
                }
                Console.WriteLine($"  Array created successfully");

                // Test Optimized Bubble Sort
                Console.WriteLine($"  Starting Optimized Bubble Sort for {size} elements...");
                stopwatch.Restart();
                BubbleSortOptimized(arr);
                stopwatch.Stop();
                Console.WriteLine($"  Sorting completed for {size} elements!");

                Console.WriteLine($"  Optimized Bubble Sort Time: {stopwatch.ElapsedMilliseconds} ms");
                Console.WriteLine($"  Optimized Bubble Sort Comparisons: {totalComparisons:N0}");
                Console.WriteLine($"  Optimized Bubble Sort Swaps: {totalSwaps:N0}");
                Console.WriteLine($"  Time per element: {(double)stopwatch.ElapsedMilliseconds / size:F4} ms");
                Console.WriteLine();
            }
        }

        private static void CompareAllBubbleSortVariants()
        {
            Console.WriteLine("=== COMPARING ALL BUBBLE SORT VARIANTS ===");
            Console.WriteLine("This demonstrates performance differences between bubble sort optimizations");
            Console.WriteLine();

            var random = new Random(RANDOM_SEED);
            var stopwatch = new Stopwatch();

            Console.WriteLine($"Creating base array with {ARRAY_SIZE} random elements...");
            // Create array with random data
            int[] arr = new int[ARRAY_SIZE];
            for (int j = 0; j < ARRAY_SIZE; j++)
            {
                arr[j] = random.Next(1, 10000);
            }
            Console.WriteLine("Base array created successfully!");

            Console.WriteLine();
            Console.WriteLine("Testing Optimized Bubble Sort...");
            // Test Optimized Bubble Sort
            int[] arr1 = (int[])arr.Clone();
            stopwatch.Restart();
            BubbleSortOptimized(arr1);
            stopwatch.Stop();
            long optimizedTime = stopwatch.ElapsedMilliseconds;
            long optimizedComparisons = totalComparisons;
            long optimizedSwaps = totalSwaps;
            Console.WriteLine("Optimized Bubble Sort completed!");

            Console.WriteLine();
            Console.WriteLine("Testing Cocktail Sort...");
            // Test Cocktail Sort
            int[] arr2 = (int[])arr.Clone();
            stopwatch.Restart();
            CocktailSort(arr2);
            stopwatch.Stop();
            long cocktailTime = stopwatch.ElapsedMilliseconds;
            long cocktailComparisons = totalComparisons;
            long cocktailSwaps = totalSwaps;
            Console.WriteLine("Cocktail Sort completed!");

            Console.WriteLine();
            Console.WriteLine("Testing Bubble Sort with Reduced Comparisons...");
            // Test Bubble Sort with Reduced Comparisons
            int[] arr3 = (int[])arr.Clone();
            stopwatch.Restart();
            BubbleSortReducedComparisons(arr3);
            stopwatch.Stop();
            long reducedTime = stopwatch.ElapsedMilliseconds;
            long reducedComparisons = totalComparisons;
            long reducedSwaps = totalSwaps;
            Console.WriteLine("Bubble Sort with Reduced Comparisons completed!");

            Console.WriteLine();
            Console.WriteLine($"Bubble Sort Variants Comparison Results:");
            Console.WriteLine($"Array size: {ARRAY_SIZE}");
            Console.WriteLine();
            Console.WriteLine($"Optimized Bubble Sort:");
            Console.WriteLine($"  Time: {optimizedTime} ms");
            Console.WriteLine($"  Comparisons: {optimizedComparisons:N0}");
            Console.WriteLine($"  Swaps: {optimizedSwaps:N0}");
            Console.WriteLine();
            Console.WriteLine($"Cocktail Sort:");
            Console.WriteLine($"  Time: {cocktailTime} ms");
            Console.WriteLine($"  Comparisons: {cocktailComparisons:N0}");
            Console.WriteLine($"  Swaps: {cocktailSwaps:N0}");
            Console.WriteLine();
            Console.WriteLine($"Bubble Sort with Reduced Comparisons:");
            Console.WriteLine($"  Time: {reducedTime} ms");
            Console.WriteLine($"  Comparisons: {reducedComparisons:N0}");
            Console.WriteLine($"  Swaps: {reducedSwaps:N0}");
            Console.WriteLine();
            Console.WriteLine($"Performance Comparison:");
            Console.WriteLine($"  Cocktail vs Optimized: {(double)optimizedTime / cocktailTime:F2}x");
            Console.WriteLine($"  Reduced Comparisons vs Optimized: {(double)optimizedTime / reducedTime:F2}x");
            Console.WriteLine();
        }

        static void Main(string[] args)
        {
            Console.WriteLine("=== OPTIMIZED BUBBLE SORT PERFORMANCE SOLUTION ===");
            Console.WriteLine("This program demonstrates optimized Bubble Sort implementations:");
            Console.WriteLine("1. Optimized Bubble Sort with early termination");
            Console.WriteLine("2. Cocktail Sort (bidirectional bubble sort)");
            Console.WriteLine("3. Bubble Sort with reduced comparisons");
            Console.WriteLine("4. Performance comparison between bubble sort variants");
            Console.WriteLine();
            Console.WriteLine($"Array size: {ARRAY_SIZE} elements");
            Console.WriteLine("This will demonstrate bubble sort optimization improvements!");
            Console.WriteLine();

            // Test different optimized bubble sort algorithms
            TestOptimizedBubbleSort(TEST_ITERATIONS);
            TestCocktailSort(TEST_ITERATIONS);
            TestBubbleSortReducedComparisons(TEST_ITERATIONS);
            TestWithDifferentArraySizes();
            CompareAllBubbleSortVariants();

            Console.WriteLine("=== OVERALL OPTIMIZATION ANALYSIS ===");
            Console.WriteLine("1. Run this with Visual Studio Profiler in INSTRUMENTATION mode");
            Console.WriteLine("2. Compare with the inefficient version to see performance improvements!");
            Console.WriteLine("3. Observe the reduction in comparisons and swaps");
            Console.WriteLine("4. Analyze the efficiency of optimized bubble sort algorithms");
            Console.WriteLine("5. Examine bubble sort optimization techniques");
            Console.WriteLine("6. Look for optimization techniques in action");
            Console.WriteLine("7. Focus on 'Hot Paths' - most frequently called functions");
            Console.WriteLine("8. Check for improved bubble sort patterns");
            Console.WriteLine();
            Console.WriteLine("Key Bubble Sort Optimization Techniques Demonstrated:");
            Console.WriteLine("- Early termination: Stop when no swaps occur in Bubble Sort");
            Console.WriteLine("- Bidirectional sorting: Cocktail sort reduces passes needed");
            Console.WriteLine("- Reduced comparisons: Skip already sorted elements");
            Console.WriteLine("- Efficient swapping: Optimize swap operations");
            Console.WriteLine("- Time complexity: Still O(n²) but with better constants");
            Console.WriteLine("- Space complexity optimization: Efficient memory usage");
            Console.WriteLine("- Reduced comparisons: Minimize unnecessary comparisons");
            Console.WriteLine("- Reduced swaps: Minimize unnecessary swaps");
            Console.WriteLine($"- Array size tested: {ARRAY_SIZE} elements");

            Console.WriteLine("Program completed successfully!");
            Console.WriteLine("Exiting now...");
            
            // Exit immediately to stop profiler timing
            return;
        }
    }
}

File: module2/class4/example1-m1p4e1-bubble-sort-optimized.cpp
/*
 * PROFILING EXAMPLE: Optimized Bubble Sort Performance Solution
 * 
 * This example demonstrates optimized sorting implementations:
 * - QuickSort with O(n log n) average time complexity
 * - MergeSort with O(n log n) guaranteed time complexity
 * - Optimized Bubble Sort with early termination
 * - Efficient comparison and swap operations
 * 
 * OBJECTIVES:
 * - Demonstrate optimization techniques for sorting algorithms
 * - Show performance improvements through better algorithms
 * - Compare inefficient vs optimized sorting solutions
 * - Identify best practices for sorting algorithm design
 * - Prepare reflection on algorithm optimization
 * 
 * NOTE: This code demonstrates optimized sorting implementations.
 * Run this with Visual Studio Profiler in Instrumentation mode
 * to observe performance improvements and optimization patterns.
 */

#include <iostream>
#include <vector>
#include <chrono>
#include <random>
#include <algorithm>

using namespace std;
using namespace std::chrono;

// ============================================================================
// CLASSROOM CONFIGURATION - EASY TO ADJUST FOR DIFFERENT DEMONSTRATIONS
// ============================================================================

// Array Configuration
const int ARRAY_SIZE = 5000;              // Array size for profiling (same as problem version)
const int TEST_ITERATIONS = 1;            // Number of test iterations
const int RANDOM_SEED = 42;               // Random seed for reproducible results

// Performance Tracking
long long totalComparisons = 0;
long long totalSwaps = 0;

// ============================================================================

/*
 * SCENARIO 1: Optimized Bubble Sort Implementation
 * Demonstrates early termination and efficiency improvements
 */

// OPTIMIZED: Bubble Sort with early termination - O(n²) worst case, O(n) best case
void bubbleSortOptimized(vector<int>& arr) {
    int n = arr.size();
    totalComparisons = 0;
    totalSwaps = 0;

    // OPTIMIZED: Early termination when no swaps occur
    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;

        for (int j = 0; j < n - i - 1; j++) {
            totalComparisons++;

            // OPTIMIZED: Efficient comparison and swap
            if (arr[j] > arr[j + 1]) {
                // OPTIMIZED: Efficient swap operation
                swap(arr[j], arr[j + 1]);
                totalSwaps++;
                swapped = true;
            }
        }

        // OPTIMIZED: Early termination if no swaps occurred
        if (!swapped) {
            break;
        }
    }
}

/*
 * SCENARIO 2: QuickSort Implementation
 * Demonstrates O(n log n) average time complexity
 */

// OPTIMIZED: QuickSort with O(n log n) average time complexity
void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        // OPTIMIZED: Partition the array
        int pivotIndex = partition(arr, low, high);

        // OPTIMIZED: Recursively sort subarrays
        quickSort(arr, low, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, high);
    }
}

// OPTIMIZED: Efficient partitioning for QuickSort
int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
        totalComparisons++;
        if (arr[j] <= pivot) {
            i++;
            swap(arr[i], arr[j]);
            totalSwaps++;
        }
    }

    swap(arr[i + 1], arr[high]);
    totalSwaps++;
    return i + 1;
}

/*
 * SCENARIO 3: MergeSort Implementation
 * Demonstrates O(n log n) guaranteed time complexity
 */

// OPTIMIZED: MergeSort with O(n log n) guaranteed time complexity
void mergeSort(vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;

        // OPTIMIZED: Recursively sort both halves
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        // OPTIMIZED: Merge the sorted halves
        merge(arr, left, mid, right);
    }
}

// OPTIMIZED: Efficient merging for MergeSort
void merge(vector<int>& arr, int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    // OPTIMIZED: Create temporary vectors
    vector<int> leftArr(n1);
    vector<int> rightArr(n2);

    // OPTIMIZED: Copy data to temporary vectors
    for (int i = 0; i < n1; i++) {
        leftArr[i] = arr[left + i];
    }
    for (int j = 0; j < n2; j++) {
        rightArr[j] = arr[mid + 1 + j];
    }

    // OPTIMIZED: Merge the temporary vectors back
    int leftIndex = 0, rightIndex = 0;
    int mergedIndex = left;

    while (leftIndex < n1 && rightIndex < n2) {
        totalComparisons++;
        if (leftArr[leftIndex] <= rightArr[rightIndex]) {
            arr[mergedIndex] = leftArr[leftIndex];
            leftIndex++;
        } else {
            arr[mergedIndex] = rightArr[rightIndex];
            rightIndex++;
        }
        mergedIndex++;
    }

    // OPTIMIZED: Copy remaining elements
    while (leftIndex < n1) {
        arr[mergedIndex] = leftArr[leftIndex];
        leftIndex++;
        mergedIndex++;
    }

    while (rightIndex < n2) {
        arr[mergedIndex] = rightArr[rightIndex];
        rightIndex++;
        mergedIndex++;
    }
}

/*
 * SCENARIO 4: Performance Testing Functions
 */

void testOptimizedBubbleSort(int iterations) {
    cout << "=== TESTING OPTIMIZED BUBBLE SORT ===" << endl;
    cout << "This demonstrates early termination optimization" << endl;
    cout << "Array size: " << ARRAY_SIZE << endl;
    cout << "Iterations: " << iterations << endl;
    cout << endl;

    random_device rd;
    mt19937 gen(RANDOM_SEED);
    uniform_int_distribution<int> dis(1, 10000);

    for (int i = 0; i < iterations; i++) {
        cout << "Testing Optimized Bubble Sort (iteration " << (i + 1) << ")..." << endl;

        // Create vector with random data
        vector<int> arr(ARRAY_SIZE);
        for (int j = 0; j < ARRAY_SIZE; j++) {
            arr[j] = dis(gen);
        }

        // Test optimized Bubble Sort
        auto start = high_resolution_clock::now();
        bubbleSortOptimized(arr);
        auto end = high_resolution_clock::now();
        auto duration = duration_cast<milliseconds>(end - start);

        cout << "Optimized Bubble Sort completed:" << endl;
        cout << "  Time: " << duration.count() << " ms" << endl;
        cout << "  Comparisons: " << totalComparisons << endl;
        cout << "  Swaps: " << totalSwaps << endl;
        cout << "  Array size: " << ARRAY_SIZE << endl;
        cout << endl;
    }
}

void testQuickSort(int iterations) {
    cout << "=== TESTING QUICKSORT ===" << endl;
    cout << "This demonstrates O(n log n) average time complexity" << endl;
    cout << "Array size: " << ARRAY_SIZE << endl;
    cout << "Iterations: " << iterations << endl;
    cout << endl;

    random_device rd;
    mt19937 gen(RANDOM_SEED);
    uniform_int_distribution<int> dis(1, 10000);

    for (int i = 0; i < iterations; i++) {
        cout << "Testing QuickSort (iteration " << (i + 1) << ")..." << endl;

        // Create vector with random data
        vector<int> arr(ARRAY_SIZE);
        for (int j = 0; j < ARRAY_SIZE; j++) {
            arr[j] = dis(gen);
        }

        // Test QuickSort
        auto start = high_resolution_clock::now();
        totalComparisons = 0;
        totalSwaps = 0;
        quickSort(arr, 0, arr.size() - 1);
        auto end = high_resolution_clock::now();
        auto duration = duration_cast<milliseconds>(end - start);

        cout << "QuickSort completed:" << endl;
        cout << "  Time: " << duration.count() << " ms" << endl;
        cout << "  Comparisons: " << totalComparisons << endl;
        cout << "  Swaps: " << totalSwaps << endl;
        cout << "  Array size: " << ARRAY_SIZE << endl;
        cout << endl;
    }
}

void testMergeSort(int iterations) {
    cout << "=== TESTING MERGESORT ===" << endl;
    cout << "This demonstrates O(n log n) guaranteed time complexity" << endl;
    cout << "Array size: " << ARRAY_SIZE << endl;
    cout << "Iterations: " << iterations << endl;
    cout << endl;

    random_device rd;
    mt19937 gen(RANDOM_SEED);
    uniform_int_distribution<int> dis(1, 10000);

    for (int i = 0; i < iterations; i++) {
        cout << "Testing MergeSort (iteration " << (i + 1) << ")..." << endl;

        // Create vector with random data
        vector<int> arr(ARRAY_SIZE);
        for (int j = 0; j < ARRAY_SIZE; j++) {
            arr[j] = dis(gen);
        }

        // Test MergeSort
        auto start = high_resolution_clock::now();
        totalComparisons = 0;
        totalSwaps = 0;
        mergeSort(arr, 0, arr.size() - 1);
        auto end = high_resolution_clock::now();
        auto duration = duration_cast<milliseconds>(end - start);

        cout << "MergeSort completed:" << endl;
        cout << "  Time: " << duration.count() << " ms" << endl;
        cout << "  Comparisons: " << totalComparisons << endl;
        cout << "  Swaps: " << totalSwaps << endl;
        cout << "  Array size: " << ARRAY_SIZE << endl;
        cout << endl;
    }
}

void testWithDifferentArraySizes() {
    cout << "=== TESTING OPTIMIZED ALGORITHMS WITH DIFFERENT ARRAY SIZES ===" << endl;
    cout << "This demonstrates O(n log n) complexity scaling" << endl;
    cout << endl;

    random_device rd;
    mt19937 gen(RANDOM_SEED);
    uniform_int_distribution<int> dis(1, 10000);

    vector<int> sizes = {100, 500, 1000, 2000, 3000};

    for (int size : sizes) {
        cout << "Testing with array size: " << size << endl;

        // Create vector with random data
        vector<int> arr(size);
        for (int j = 0; j < size; j++) {
            arr[j] = dis(gen);
        }

        // Test QuickSort
        auto start = high_resolution_clock::now();
        totalComparisons = 0;
        totalSwaps = 0;
        quickSort(arr, 0, arr.size() - 1);
        auto end = high_resolution_clock::now();
        auto duration = duration_cast<milliseconds>(end - start);

        cout << "  QuickSort Time: " << duration.count() << " ms" << endl;
        cout << "  QuickSort Comparisons: " << totalComparisons << endl;
        cout << "  QuickSort Swaps: " << totalSwaps << endl;
        cout << "  Time per element: " << (double)duration.count() / size << " ms" << endl;
        cout << endl;
    }
}

void compareAllAlgorithms() {
    cout << "=== COMPARING ALL SORTING ALGORITHMS ===" << endl;
    cout << "This demonstrates performance differences between algorithms" << endl;
    cout << endl;

    random_device rd;
    mt19937 gen(RANDOM_SEED);
    uniform_int_distribution<int> dis(1, 10000);

    // Create vector with random data
    vector<int> arr(ARRAY_SIZE);
    for (int j = 0; j < ARRAY_SIZE; j++) {
        arr[j] = dis(gen);
    }

    // Test Optimized Bubble Sort
    vector<int> arr1 = arr;
    auto start = high_resolution_clock::now();
    bubbleSortOptimized(arr1);
    auto end = high_resolution_clock::now();
    auto bubbleDuration = duration_cast<milliseconds>(end - start);
    long bubbleTime = bubbleDuration.count();
    long bubbleComparisons = totalComparisons;
    long bubbleSwaps = totalSwaps;

    // Test QuickSort
    vector<int> arr2 = arr;
    start = high_resolution_clock::now();
    totalComparisons = 0;
    totalSwaps = 0;
    quickSort(arr2, 0, arr2.size() - 1);
    end = high_resolution_clock::now();
    auto quickDuration = duration_cast<milliseconds>(end - start);
    long quickTime = quickDuration.count();
    long quickComparisons = totalComparisons;
    long quickSwaps = totalSwaps;

    // Test MergeSort
    vector<int> arr3 = arr;
    start = high_resolution_clock::now();
    totalComparisons = 0;
    totalSwaps = 0;
    mergeSort(arr3, 0, arr3.size() - 1);
    end = high_resolution_clock::now();
    auto mergeDuration = duration_cast<milliseconds>(end - start);
    long mergeTime = mergeDuration.count();
    long mergeComparisons = totalComparisons;
    long mergeSwaps = totalSwaps;

    cout << "Algorithm Comparison Results:" << endl;
    cout << "Array size: " << ARRAY_SIZE << endl;
    cout << endl;
    cout << "Optimized Bubble Sort:" << endl;
    cout << "  Time: " << bubbleTime << " ms" << endl;
    cout << "  Comparisons: " << bubbleComparisons << endl;
    cout << "  Swaps: " << bubbleSwaps << endl;
    cout << endl;
    cout << "QuickSort:" << endl;
    cout << "  Time: " << quickTime << " ms" << endl;
    cout << "  Comparisons: " << quickComparisons << endl;
    cout << "  Swaps: " << quickSwaps << endl;
    cout << endl;
    cout << "MergeSort:" << endl;
    cout << "  Time: " << mergeTime << " ms" << endl;
    cout << "  Comparisons: " << mergeComparisons << endl;
    cout << "  Swaps: " << mergeSwaps << endl;
    cout << endl;
    cout << "Performance Improvement:" << endl;
    cout << "  QuickSort vs Bubble Sort: " << (double)bubbleTime / quickTime << "x faster" << endl;
    cout << "  MergeSort vs Bubble Sort: " << (double)bubbleTime / mergeTime << "x faster" << endl;
    cout << endl;
}

int main() {
    cout << "=== OPTIMIZED SORTING ALGORITHMS PERFORMANCE SOLUTION ===" << endl;
    cout << "This program demonstrates optimized sorting implementations:" << endl;
    cout << "1. Optimized Bubble Sort with early termination" << endl;
    cout << "2. QuickSort with O(n log n) average time complexity" << endl;
    cout << "3. MergeSort with O(n log n) guaranteed time complexity" << endl;
    cout << "4. Performance comparison between algorithms" << endl;
    cout << endl;
    cout << "Array size: " << ARRAY_SIZE << " elements" << endl;
    cout << "This will demonstrate significant sorting performance improvements!" << endl;
    cout << endl;

    // Test different optimized algorithms
    testOptimizedBubbleSort(TEST_ITERATIONS);
    testQuickSort(TEST_ITERATIONS);
    testMergeSort(TEST_ITERATIONS);
    testWithDifferentArraySizes();
    compareAllAlgorithms();

    cout << "=== OVERALL OPTIMIZATION ANALYSIS ===" << endl;
    cout << "1. Run this with Visual Studio Profiler in INSTRUMENTATION mode" << endl;
    cout << "2. Compare with the inefficient version to see performance improvements!" << endl;
    cout << "3. Observe the dramatic reduction in comparisons and swaps" << endl;
    cout << "4. Analyze the efficiency of optimized algorithms" << endl;
    cout << "5. Examine time complexity improvements" << endl;
    cout << "6. Look for optimization techniques in action" << endl;
    cout << "7. Focus on 'Hot Paths' - most frequently called functions" << endl;
    cout << "8. Check for improved time complexity patterns" << endl;
    cout << endl;
    cout << "Key Sorting Optimization Techniques Demonstrated:" << endl;
    cout << "- Early termination: Stop when no swaps occur in Bubble Sort" << endl;
    cout << "- Better algorithms: Use O(n log n) algorithms instead of O(n²)" << endl;
    cout << "- Efficient partitioning: Optimize pivot selection in QuickSort" << endl;
    cout << "- Efficient merging: Optimize merge operations in MergeSort" << endl;
    cout << "- Time complexity improvement: O(n²) -> O(n log n)" << endl;
    cout << "- Space complexity optimization: Efficient memory usage" << endl;
    cout << "- Reduced comparisons: Minimize unnecessary comparisons" << endl;
    cout << "- Reduced swaps: Minimize unnecessary swaps" << endl;
    cout << "- Array size tested: " << ARRAY_SIZE << " elements" << endl;

    return 0;
}
